nmagent/client.go

```diff
@@ -0,0 +1,208 @@
+package nmagent
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/Azure/azure-container-networking/nmagent/internal"
+	"github.com/pkg/errors"
+)
+
+// NewClient returns an initialized Client using the provided configuration
+func NewClient(c Config) (*Client, error) {
+	if err := c.Validate(); err != nil {
+		return nil, errors.Wrap(err, "validating config")
+	}
+
+	client := &Client{
+		httpClient: &http.Client{
+			Transport: &internal.WireserverTransport{
+				Transport: http.DefaultTransport,
+			},
+		},
+		host:      c.Host,
+		port:      c.Port,
+		enableTLS: c.UseTLS,
+		retrier: internal.Retrier{
+			// nolint:gomnd // the base parameter is explained in the function
+			Cooldown: internal.Exponential(1*time.Second, 2),
+		},
+	}
+
+	return client, nil
+}
+
+// Client is an agent for exchanging information with NMAgent
+type Client struct {
+	httpClient *http.Client
+
+	// config
+	host string
+	port uint16
+
+	enableTLS bool
+
+	retrier interface {
+		Do(context.Context, func() error) error
+	}
+}
+
+// Option is a functional option for configuration optional behavior in the
+// client
+type Option func(*Client)
+
+// JoinNetwork joins a node to a customer's virtual network
+func (c *Client) JoinNetwork(ctx context.Context, jnr JoinNetworkRequest) error {
+	req, err := c.buildRequest(ctx, jnr)
+	if err != nil {
+		return errors.Wrap(err, "building request")
+	}
+
+	err = c.retrier.Do(ctx, func() error {
+		resp, err := c.httpClient.Do(req) // nolint:govet // the shadow is intentional
+		if err != nil {
+			return errors.Wrap(err, "executing request")
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			return c.error(resp.StatusCode, resp.Header, resp.Body)
+		}
+		return nil
+	})
+
+	return err // nolint:wrapcheck // wrapping this just introduces noise
+}
+
+// GetNetworkConfiguration retrieves the configuration of a customer's virtual
+// network. Only subnets which have been delegated will be returned
```

Comment:
My understanding is that all subnets in vnet are returned, not just delegated ones. We mark the subnets as delegated at later steps in DNC.

Reply:
I pulled most of these doc comments straight from the NMA docs just so users wouldn't have to cross reference them all the time... This one comes from section 8.2.6. We should find out if that's true and update the upstream docs if all the subnets are returned.